### ISSUE TITLE
Catmull-Rom Splines

### ISSUE THREAD

**Issue by renpytom** [renpy/renpy#1687]
https://lemmasoft.renai.us/forums/viewtopic.php?p=503020#p503020

**Comment by RenpyRemix** [comment]
A simplified version that just has two functions... (might be easier to implement)

Note: It still supports alpha (curvature changes) and N-dimension nodes though does not linearize (equidistant) the points over time, it just calculates a position at a requested time...

Included below the functions is a simple screen and start label to show the curves and the spacing of points even though the time progress is linear...

Hope this helps

```


init python:

    def get_pos_for_time(nodes=[], t=0.0, alpha=0.5):
        """
        Return a position (tuple or list) to represent 
        the co-ordinates for the passed time t which is
        a float between 0.0 and 1.0 reflecting the full
        movement between nodes[1] and nodes[-2]
        """
        if not isinstance(t, float) or t < 0.0 or t > 1.0:
            raise ValueError, "t should be a float in range 0.0 to 1.0"

        if len(nodes) < 4:
            raise ValueError, "Catmull Rom requires at least 4 nodes"

        inner_node_count = float(len(nodes) - 3)

        # determine which nodes are relevant
        sector = int(t // ( 1.0 / inner_node_count ) + 1)

        # determine t for this sector
        st = ( t % ( 1.0 / inner_node_count ) ) * inner_node_count

        if t == 1.0:

            sector, st = sector - 1, 1.0

        return get_point(st, *nodes[sector-1:sector+3], alpha=alpha)


    def get_point(t, p0, p1, p2, p3, alpha=0.5):
        """
        Calculate a point between p1 and p2
        using p0 and p3 as control points
        where t is a float % between p1 and p2
        """

        def alpha_time(ti, Pi, Pj):
            return ( sum( [ (Pj[n]-Pi[n])**2 
                            for n in range(len(Pi)) ] )**0.5 )**alpha + ti

        t0 = 0
        t1 = alpha_time(t0, p0, p1)
        t2 = alpha_time(t1, p1, p2)
        t3 = alpha_time(t2, p2, p3)

        t = t1 + ( (t2 - t1) * t )
        pts_a = []
        for x in range(len(p1)):
            a1 = (t1-t) / (t1-t0) * p0[x] + (t-t0) / (t1-t0) * p1[x]
            a2 = (t2-t) / (t2-t1) * p1[x] + (t-t1) / (t2-t1) * p2[x]
            a3 = (t3-t) / (t3-t2) * p2[x] + (t-t2) / (t3-t2) * p3[x]

            b1 = (t2-t) / (t2-t0) * a1 + (t-t0) / (t2-t0) * a2
            b2 = (t3-t) / (t3-t1) * a2 + (t-t1) / (t3-t1) * a3

            pts_a.append(
                (t2-t) / (t2-t1) * b1 + (t-t1) / (t2-t1) * b2
            )
        return type(p0)(pts_a)



screen catmull_showcase(*args, **kwargs):

    fixed:

        for k in range(0, 101):

            $ pt = get_pos_for_time(
                nodes = args,
                t = float(k) / 100.0,
                alpha = kwargs.get('alpha', 0.5) )


            add Solid("#5F5A", xysize=(3,3)):
                pos ( int(round(pt[0]))-1,
                      int(round(pt[1]))-1 )
                if len(pt) > 2:
                    alpha pt[2]

        for i,arg in enumerate(args):

            add Text(str(i), color="#FFFA"):
                anchor (0.5, 0.5)
                pos ( int(round(arg[0])),
                      int(round(arg[1])) )


label start:

    scene expression Solid("000")

    "Start..."

    show screen catmull_showcase(
        [ 300,300 ],
        [ 250,315 ],
        [ 450,500 ],
        [ 600,250 ],
        [ 750,500 ],
        [ 900,300 ],
        alpha = 1.0 )

    "..."

    # 3 dimension nodes

    show screen catmull_showcase(
        [ 300,100,0.3 ],
        [ 450,200,0.7 ],
        [ 550,150,0.9 ],
        [ 600,400,0.4 ],
        [ 150,300,0.6 ],
        [ 300,100,1.0 ],
        alpha = 0.5 )

    "..."

    return
```

**Comment by Andykl** [comment]
Implemented in #2103
